### PR TITLE
Add scatter_add op

### PR DESCRIPTION
**Pull request by thisjiang**
As title. Add `scatter_add` op, the op like as [`torch.scatter_add`](https://pytorch.org/docs/stable/generated/torch.Tensor.scatter_add_.html#torch.Tensor.scatter_add_),  but the `index` only supporting 1-dimension. In other words, the `cinn.scatter_add` more like the `cinn.index_assign` with add (Or we can say `index_add`).

```
builder.scatter_add(x, updates, index, axis)
```
where :

1.`len(index.shape) == 1`
2. `len(updates.shape) == len(x.shape)`
3.`updates.shape[i] == x.shape[i] if i != axis` 
4. `updates.shape[axis] == len(index)`

For example:
```
x = [[1, 2, 3], [4, 5, 6], [7, 8, 9]] -> shape[3, 3]
axis = 1
index = [0, 2] ->len[2]
updates = [[1, 3], [4, 6], [7, 9]] ->shape[3, 2]

output=[[2, 2, 6], [8, 5, 12], [14, 8, 18]] ->shape[3, 3]
```

### CODE REVIEW

File: cinn/runtime/cuda/cinn_cuda_runtime_source.cuh [hunk 1]
```diff
@@ -178,3 +178,37 @@ __device__ inline int cinn_cuda_find_float(const float *buf, int size, float num
...
+      if (buf[i] == num) return i;                          \
+    }                                                       \
+    return -1;                                              \
+  } while(0)
+
+__device__ inline int cinn_cuda_find_int_from(const int *buf, int size, int num, int begin) {
+  __cinn_cuda_find_from_kernel(buf, size, num, begin);
+}
+
+__device__ inline int cinn_cuda_find_float_from(const float *buf, int size, float num, int begin) {
```
Comment (SunNy820828449) [inline cinn/runtime/cuda/cinn_cuda_runtime_source.cuh:194]: 有用到这个函数吗？

Comment (thisjiang) [reply cinn/runtime/cuda/cinn_cuda_runtime_source.cuh:194]: 暂时还没有用到。。。

File: cinn/hlir/pe/transform.cc [hunk 2]
```diff
@@ -945,6 +945,46 @@ ir::Tensor IndexAssign(const ir::Tensor& input,
...
+
+  auto pos_axis = axis;
+  if (pos_axis < 0) pos_axis += input->shape.size();
+  CHECK(pos_axis >= 0 && pos_axis < input->shape.size())
+      << "Param [axis] of IndexAdd should satisfy 0 <= axis < input.shape ! Please Check.\n";
+
+  std::vector<int> strides(updates->shape.size(), 1);
+  for (int i = updates->shape.size() - 2; i >= 0; --i) {
+    strides[i] = strides[i + 1] * updates->shape[i + 1].as_int32();
+  }
```
Comment (wzzju) [inline cinn/hlir/pe/transform.cc:969]: 这部分请加一些注释，解释strides的含义。

File: cinn/hlir/pe/transform.cc [hunk 3]
```diff
@@ -945,6 +945,46 @@ ir::Tensor IndexAssign(const ir::Tensor& input,
...
+
+  auto output = Compute(
```
Comment (wzzju) [inline cinn/hlir/pe/transform.cc:989]: 请在这里大致说明一下Compute函数所执行的操作含义。

File: cinn/hlir/pe/transform.cc (hunk 2 above)
Comment (thisjiang) [reply cinn/hlir/pe/transform.cc:969]: Done

File: cinn/hlir/pe/transform.cc (hunk 3 above)
Comment (thisjiang) [reply cinn/hlir/pe/transform.cc:989]: Done

**Review by wzzju — APPROVED** [review]
LGTM

**Pull request merged**